tiledb/api/proxy/proxy.h

```diff
@@ -0,0 +1,917 @@
+   *    processes the event. This results in one single cycle.
+   * 2. The queue is not empty and the event is non-blocking. The message is
+   *    enqueued. This event moves from `Receive` to `Quiet` and returns.
+   * 3. The queue is not empty and the event is blocking. The message is
+   *    enqueued and blocks waiting to move to 'Hand-off'. The loop continues
+   *    until the queue is empty.
+   *
+   * Put more simply, the `event()` calls for blocking events are able to handle
+   * queued non-blocking ones.
+   *
+   * This behavior is necessary to support prompt shutdown. The `shutdown` event
+   * is non-blocking so that a top-level shutdown thread is not at risk of
+   * blocking on any potentially long-lived state transitions. The `construct`
+   * event is considered long lived because it may involve storage I/O
+   * operations.
+   */
+
+  /**
+   * Perform a state transition and return holding a lock that prevents other
+   * state transitions.
+   *
+   * This function blocks if the event cannot be processed immediately. A block
+   * may happen either in by the queue mutex or the state machine mutex.
+   *
+   * @pre `lock_state` does not hold a lock on the state mutex.
+   *
+   * @param lock A lock object to be used for the state machine
+   */
+  void event(lock_type& lock_state, ProxyEvent ev) {
+    // Assert: event state is `Quiet`
+    /*
+     * Create the queue lock always in the locked state. Queue processing has
+     * nothing long-lived, so we always block for exclusive access to the queue.
+     */
+    std::unique_lock<std::mutex> lock_queue{m_queue_};
+    // Assert: event state is `Receive`
+
+    /*
+     * Try the transition to `Hand-off`. If it doesn't immediately succeed and
+     * we have a non-blocking event, queue the event and return.
+     */
+    if (lock_state.try_lock()) {
+      // We have a lock on the state mutex
+      // Assert: event state is `Hand-off`
+    } else {
+      // We do not have a lock on the state mutex
+      if (will_block[to_index(ev)]) {
+        /*
+         * `lock()` blocks until it obtains a lock.
+         */
+        lock_state.lock();
+        // Assert: event state is `Hand-off`
+      } else {
+        /*
+         * We cannot immediately process a non-blocking event, so we
+         * queue it up and leave.
+         */
+        queue_.push(ev);
+        return;
+        /*
+         * Assert: event state is `Quiet` after the destructor for
+         * `lock_queue` executes
+         */
+      }
+      // Assert: event state is `Hand-off`
+      // In we had a non-blocking event, this function already returned.
+    }
+    // Assert: event state is `Hand-off`
+    // Whether `try_lock` locked or not, we now have a lock.
+
+    ProxyEvent event;
+    if (queue_.empty()) {
+      event = ev;
+    } else {
+      queue_.push(ev);
+      event = queue_.front();
+      queue_.pop();
+    }
+    // Assert: `event` is initialized
```

Comment:
I know `event` isn't technically uninitialized here, and it isn't const, but what are your thoughts on this kind of idiom:
```c++
    ProxyEvent event{[this, &ev]() {
      if (queue_.empty()) {
        return ev;
      } else {
        queue_.push(ev);
        auto event = queue_.front();
        queue_.pop();
        return event;
      }
    }()};
```